controllers/dspipeline_controller.go

```diff
@@ -0,0 +1,265 @@
+		log.Error(err, "Unable to fetch the DSPipelineParams")
+		return ctrl.Result{}, err
+	}
+
+	err = params.ExtractParams(dspipeline)
+	if err != nil {
+		log.Error(err, "Unable to parse CR spec, "+
+			"failed to reconcile, ensure CR is well formed")
+		return ctrl.Result{}, err
+	}
+
+	if dspipeline.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(dspipeline, finalizerName) {
+			controllerutil.AddFinalizer(dspipeline, finalizerName)
+			if err := r.Update(ctx, dspipeline); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(dspipeline, finalizerName) {
+			if err := r.cleanUpResources(dspipeline, ctx, req, params); err != nil {
+				return ctrl.Result{}, err
+			}
+			controllerutil.RemoveFinalizer(dspipeline, finalizerName)
+			if err := r.Update(ctx, dspipeline); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		// Stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+	}
+
+	usingCustomDB, err := params.UsingCustomDB(dspipeline)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	usingCustomStorage, err := params.UsingCustomStorage(dspipeline)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if !usingCustomStorage {
+		err := r.ReconcileStorage(dspipeline, ctx, req, params)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	if !usingCustomDB {
+		err = r.ReconcileDatabase(dspipeline, ctx, req, params)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// TODO: Ensure db/storage (if deploying custom) are running before
+	// Use status fields to conditionally deploy
+
+	err = r.ReconcileAPIServer(dspipeline, ctx, req, params)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.ReconcilePersistenceAgent(dspipeline, params)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.ReconcileScheduledWorkflow(dspipeline, params)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.ReconcileUI(dspipeline, params)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.ReconcileViewerCRD(dspipeline, params)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// TODO: After the build completes, image field is replaced with sha tag
+	// Need to skip reconiliation in this case
+	//err = r.ReconcileVisualizationServer(dspipeline, ctx, req, params)
+	//if err != nil {
+	//	return ctrl.Result{}, err
+	//}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *DSPipelineReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&dspipelinesiov1alpha1.DSPipeline{}).
+		Owns(&appsv1.Deployment{}).
+		Owns(&corev1.Secret{}).
+		Owns(&corev1.ConfigMap{}).
+		Owns(&corev1.Service{}).
+		Owns(&corev1.ServiceAccount{}).
+		Owns(&corev1.PersistentVolumeClaim{}).
+		Owns(&rbacv1.Role{}).
+		Owns(&rbacv1.RoleBinding{}).
+		Owns(&routev1.Route{}).
+		// Visualization Server Build watchers
```

Comment:
Why are these commented out right now?

Reply:
The logic mostly works, and will successfully deploy visualization server but will launch into an infinite loop once build completes because it will over-write the image field. I didn't pursuse this further since inclusion of visualizaion server didn't seem certain at the time. I can remove it if we want to just not include it all together.